### PR TITLE
Add Approaches for Collatz Conjecture

### PR DESCRIPTION
**Pull request by MatthijsBlom**

### CODE REVIEW

**Comment by MatthijsBlom** [comment]
Note: large parts of these `content.md`s are shared between Hamming (#1141) and Collatz Conjecture (here). I will do my best to incorporate in both feedback on either.

**Review by ErikSchierboom — APPROVED** [review]

**Comment by MatthijsBlom** [comment]
@petertseng Any chance you could take a look at this in the near future? There really is demand for these and other _Dig Deeper_ materials. I have been pointing individual students to these (and the other) documents manually, but that isn't terribly efficient, and likely quite a few students have been missing out.

Alternatively, would you be fine with me merging _Dig Deeper_ PRs myself? In that case I would apply to become co-maintainer.

File: exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+Sometimes, when solving a problem, it is more convenient or efficient to keep track of some kind of _state_.
+Many other languages use local variables for this.
+However, variables do not exist in Haskell: all bindings are constants.
+
+In lieu of local variables we can add extra parameters to our functions.
+These will hold our state.
+'Changing' of the state is done by calling the same functions with different arguments.
+
+Functions that uses such additional parameters to represent state are often called _workers_, and functions that use them are often called _wrappers_.
```
Comment (MatthijsBlom) [inline exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md:27]:
```suggestion
Functions that uses such additional parameters to represent state are often called _workers_, and functions that use workers to solve a problem in a stateful way are often called _wrappers_.
```

File: exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+The worker&ndash;wrapper pattern is more general than demonstrated here.
+For a few more more examples, see the [relevant wiki article][wiki-worker-wrapper], and for yet more examples and an explanation of the general pattern see the paper «[The worker/wrapper transformation][paper-worker-wrapper]» (pdf).
+
+
+## Bang patterns
+
+The above implementation of `length'` _might_ evaluate as efficiently as illustrated.
+However, it also might not!
+The definition of `length'` allows evaluation to proceed as follows as well.
```
Comment (MatthijsBlom) [inline exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md:102]:
```suggestion
The above implementation of `length'` _might_ evaluate as efficiently as illustrated.
However, depending on which optimization opportunities are spotted by the compiler, it also might not!
The definition of `length'` allows evaluation to proceed as follows as well.
```

File: exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+Still a computation as large as the list was long is built.
+The situation is not quite as bad as before though.
+
+In the case of the naive implementation of `length`, the built up expression has the form
+
+```haskell
+_ = 1 + (1 + (⋯(1 + length xs)⋯))
+```
+
+This expression cannot be simplified before `length xs` has been evaluated.
```
Comment (MatthijsBlom) [inline exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md:128]:
~~~suggestion
In the case of the naive implementation of `length`, the built up expression has the form

```haskell
_ = 1 + (1 + (⋯(1 + length …)⋯))
```

This expression cannot be simplified before `length …` has been evaluated.
~~~

**Review by MatthijsBlom — COMMENTED** [review]
@NgWenYang I have proposed some edits. Do these resolve your confusion?

**Comment by ErikSchierboom** [comment]
@MatthijsBlom I'm happy to merge this and to improve things in a follow-up PR. It looks like the suggestions are non-breaking.

**Comment by MatthijsBlom** [comment]
@ErikSchierboom Fine by me. Let's. Does the same apply to [the other Dig Deeper PRs](https://github.com/exercism/haskell/pulls?q=is%3Apr+is%3Aopen+%22Add+Approaches%22)?

**Comment by ErikSchierboom** [comment]
Yep. Done (except for the draft ones)

**Pull request merged**